0012-common-ledger-api/0012-common-ledger-api.md

```diff
@@ -0,0 +1,1201 @@
+
+Some things not included in this specification:
+
+- Account creation and management
+- Permissions management and delegation
+- A way to increase or decrease the net sum of balances in the ledger
+- Transfers that execute automatically, dependent on other transfers
+- Transfers that can credit or debit more than one account atomically
+
+For a fully-functional reference implementation, see the [five-bells-ledger](https://github.com/interledgerjs/five-bells-ledger).
+
+
+### Authorization and Authentication
+
+The ledger MUST authenticate API clients using any of the following methods:
+
+- [HTTP Basic Auth](https://tools.ietf.org/html/rfc2617#section-2)
+- [TLS Client Certificates](https://tools.ietf.org/html/rfc5246#section-7.4.6)
+- Token-based authentication. A ledger MUST support token-based auth for [WebSocket][] connections; the ledger MAY support token-based auth for other methods. See also: [Get Auth Token][].
+
+The ledger MUST support authenticating a client as the owner of a specific account, so that it can grant or restrict access to certain methods. A ledger MAY define additional authorization levels, especially for functions that extend this API. We suggest the following authorization levels:
+
+- Administrator - Full control over all ledger operations.
+- Account owner (for a specific account) - Can perform operations relating to one specific account, such as authorizing debits from or credits to the account, and viewing the account balance.
+- Unauthorized. Read-only access to some global ledger data.
+
+
+## Data Types
+
+### Transfer Resource
+[Transfer resource]: #transfer-resource
+
+A transfer represents money being moved around _within a single ledger_. A transfer debits one account and credits another account for the exact same amount. A transfer can be conditional upon a supplied [Crypto-Condition][], in which case it executes automatically when presented with the fulfillment for the condition. (Assuming the transfer has not expired or been rejected first.) If no Crypto-Condition is specified, the transfer is unconditional, and executes as soon as it is prepared.
+
+The Common Ledger API does not define a way to add or remove money from the ledger; the methods defined here always maintain a zero sum across all account balances. However, you can effectively add money to a ledger by transferring it from an "issuer" account whose balance is allowed to go negative.
+
+A transfer object contains the fields from the following table. Some fields are _Server-provided_, meaning they cannot be set directly by clients. Fields that are "Optional" or "Server-provided" in the following table may be omitted by clients submitting transfer objects to the API, but **those fields are not optional to implement**. All fields, including the `memo` fields, must be implemented for the Common Ledger API to work properly. Fields of nested objects are indicated with a dot (`.`) character; no field names contain a dot literal.
```

Comment:
I agree on using "ledger" instead of "server."

However, I think the term "wallet" tends to be rather confusing and misleading. People use the term "wallet" to mean "client application", "key pair", or "web service" somewhat interchangeably, which is a problem. I'm going to stick with "client application" for now.